Improve create tables doc, replication doc, and cluster settings doc

norosa opened the pull request:
Closes <https://github.com/crate/crate/issues/11238>.

Specifically:

- Reworked the "Creating tables" page:

  - Added a proper introduction to the topic

  - Renamed the "Basics" section to "Table definition" and moved the "Schemas" and "Naming restrictions" under it as subsections.

  - A new admonition specifically to advertise the `column_policy` table setting (because I think dynamic columns are a key feature of CrateDB and this is relevant to table schema definition).

  - Reworded a lot of the text.

  - Improved use of RST labels and some minor RST reformatting.

  - Renamed the "Advanced use" use section to "Table configuration" and expanded the section by quite a bit to highlight what I think as some of the key features related to table settings (and table creation more generally).

    I added lots of explanation because I noticed that some of this is not really made explicit anywhere else.

- Completely reworked the "Replication" page:

  - While working on the "Creating tables" page I was trying to explain the benefits of replication and I could not find anything in the docs that explicitly talks about the fact that one of the primary benefits of replication is the fact that replica shards can be promoted to primary in the case of node loss.

    I thought this issue was important enough (and related to my other work) that I should address it immediately.

  - Reworded and expanded the intro. Reworded and moved a note about cluster health here.

  - Removed the TOC and the single "Configuration" header. There is no need to add either if there is only one section on the page.

    Reworded the rest of the content that was under this heading.

- Improvements for the "Cluster settings" page:

  - Moved "Shard balancing" to the top of the "Routing allocation" section because the rest of the subsections are closely related to each other and I think they benefit from being ordered sequentially.

  - Created a new subsection called "Attribute-based shard allocation" that groups together the next two sections:

    - "Cluster-wide attribute awareness" (renamed from "Awareness")
    - "Cluster-wide attribute filtering" (renamed from "Cluster-wide allocation filtering")

    My hope for the renaming is that overall it makes it more clear what these things do (putting custom attributes to work) and how they relate to each other (see next item).


  - Added intros to all three sections mentioned above.

    I wanted to better advertise the custom attribute functionality. Prior to this change:

    - There was no *clear* connection made between custom attributes and "awareness" or "allocation filtering" making it difficult to understand what the latter two configuration options were for.

    - It was difficult to grasp what custom attributes are useful for.

    While not specifically related to table settings, I did this because I think this is a key feature of CrateDB that deserves better highlighting.

- Other files:

    - Minor updates to reflect renamed RST labels

norosa commented:
@seut I had to guess at some of the explanations I added so they might not be quite right so I tagged you so you can give it some close scrutiny 😅

seut reviewed (commented):
Nice work, added a couple of comments.

norosa commented on docs/general/ddl/create-table.rst:
```diff
@@ -1,47 +1,73 @@
+    the ``CREATE TABLE`` statement (what's known as a *strict* :ref:`column
+    policy <column_policy>`). However, you can configure the
+    :ref:`column_policy table parameter <column_policy>` to allow the dynamic
+    creation of new columns at query-time (what's known as a *dynamic* column
```
changed to:

> By default, CrateDB will enforce the column definitions you specified with the CREATE TABLE statement (what’s known as a strict column policy).
>
> However, you can configure the column_policy table parameter so that the INSERT, UPDATE, and COPY FROM statements can arbitrarily create new columns as needed (what’s known as a dynamic column policy).

norosa commented on docs/general/ddl/create-table.rst:
```diff
@@ -147,15 +174,78 @@ Column names are restricted in terms of patterns:
+
+      `How-to guides: Tuning partitions for insert performance`_
+
+- You can split partitions into one or more :ref:`shards <ddl-sharding>` with
```
hmm. the doc on sharding says (first sentence): 

> Every table partition is split into a configured number of shards.

my reading of this was that every table technically has "one" partition, even if you didn't explicitly configure it. this is also the reason I put this item second because I figured there was a hierarchy of concepts: table -> partition -> shard

but I guess from your comment, even if that is technically the case, perhaps it's more of a technicality than a useful thing to articulate in this context?

anyway, I moved it to the top and reworded it:

> CrateDB transparently segments the underlying storage of table data into shards (four by default). You can configure the number of shards with the CLUSTERED BY clause. You control how CrateDB routes table rows to shards by specifying a routing column.

side note:

this was actually the paragraph that sent me down a bit of a rabbit hole on Monday. it's really hard to describe what a shard is in simple English without a lot of text to introduce all the concepts needed to understand everything if it's entirely new to you. the two biggest problems being:

- making it clear to the reader that shards are not something that affects your typical read and write interactions with the table
- being clear when describing what a shard actually is on disk in a way that a) is technically accurate and doesn't introduce confusion with regards to the specifics of CrateDB's implementation, Elasticsearch, Lucene, etc., and b) communicates its usefulness in terms of being a "thing" at the filesystem level that can "move" (or rather, be transferred) between nodes

the sentence in my rewritten paragraph makes some assumptions re the reader's preexisting understanding that I'm not satisfied with, but it will do for now. I'd like to come back this when I do some more glossary work

norosa replied:
I moved the replication bullet point so it appears after this one also

makes sense to keep them sequential imo

norosa commented on docs/general/ddl/create-table.rst:
```diff
@@ -147,15 +174,78 @@ Column names are restricted in terms of patterns:
+  <conf-routing-allocation-disk>`, or both).
+
+  By :ref:`distributing shards <concept-clustering>` across the cluster, you
+  can increase both `data availability`_ and `service resilience`_. With
```
my thinking for this was that if a node holding one shard splits off from the cluster, you can still read the table, but some data will be missing. is that correct? if correct, my reasoning is:

- availability is improved in the case of unreliable networks and node failures (both pretty much guaranteed to happen at one point or another)
- this also makes the cluster more resilient to failures
- more shards on more nodes mean more opportunities for distributed query execution

the same would be true of replicas

if my understanding is correct and my reasoning is accurate (and useful for the reader), perhaps it would make sense to keep this para here, and then in the next bullet point about replication, I could make it more explicit that more replicas offer the same benefit?

related question:

- I added the point about read performance to the replicas bullet because I only saw that mentioned when reviewing the docs on replicas. do more shards (primary) also improve read performance? I suppose the would if "read performance" is a proxy for "increased parallelization via distributed query execution"

seut replied:
> my thinking for this was that if a node holding one shard splits off from the cluster, you can still read the table, but some data will be missing. is that correct?

No that's not correct, the queries will fail.

> do more shards (primary) also improve read performance?

Not in general, it depends. E.g. multiple shards increases parallel execution but also requires merge phases. So using multiple shards on one node with multiple cpu's/core's probably perform better than using a single shard. With multiple shards across multiple nodes more variables come into account like more merges, network traffic, etc.

norosa replied:
aha. gotcha. thanks. I removed this para entirely

in the next bullet, on replication, I changed the final para to this:

> Replication can also improve read performance when it results in copies of the data being available on multiple nodes because this gives CrateDB more opportunities to parallelize query execution.

have I understood this (cause and effect) correctly?

norosa commented on docs/general/ddl/replication.rst:
```diff
@@ -1,73 +1,85 @@
-stored additionally on so called secondary shards. This can be useful for
-better read performance and high availability.
+You can configure CrateDB to *replicate* tables. When a table is replicated,
+table data is stored on multiple primary shards, with each primary shard having
```
I have reworked this entire intro. see next comment

norosa commented on docs/general/ddl/replication.rst:
```diff
@@ -1,73 +1,85 @@
+Replica shards can also improve read performance because more copies of the
+data spread across more nodes. Having more copies of the data on more nodes can
+also increases the opportunities for CrateDB to `parallelize`_ query
+execution.
```
based on this and other comments you made, I reworded and expanded the intro:

> You can configure CrateDB to replicate tables. When you configure replication, CrateDB will try to ensure that every table shard has one or more available copies at all times.
>
> When there are multiple copies of the same shard, CrateDB will mark one as the primary shard and treat the rest as replica shards. Write operations always go to the primary shard, whereas read operations can go to any shard. CrateDB continually synchronizes data from the primary shard to all replica shards (through a process known as shard recovery).
>
> When a primary shard is lost (e.g., due to node failure), CrateDB will promote a replica shard to primary. Hence, more table replicas mean a smaller chance of permanent data loss (through increased data redundancy) in exchange for more disk space utilization and intra-cluster network traffic.
> 
> CrateDB allocates every shard to a specific node.CrateDB will try to allocate the primary shard and replica shards to different nodes to maximize durability and resiliency.
>
> Replication can also improve read performance because any increase in the number of shards distributed across a cluster increases the opportunities for CrateDB to parallelize query execution across multiple nodes.

norosa commented on docs/general/ddl/replication.rst:
```diff
@@ -1,73 +1,85 @@
+    The `CrateDB Admin UI`_ provides visual indicators of cluster health that
+    take replication status into account. Alternatively, you can query health
+    information for yourself directly from the :ref:`sys.health <sys-health>`
+    table.
```
now reads:

> The CrateDB Admin UI provides visual indicators of cluster health that take replication status into account. Alternatively, you can query health information for yourself directly from the sys.health table. Detailed replication information can be queried from the sys.shards and sys.allocations tables.

norosa commented on docs/general/ddl/replication.rst:
```diff
@@ -1,73 +1,85 @@
+Range     Explanation
+========= =====================================================================
+``0-1``   If you only have one node, CrateDB will not create any replicas.
+          Having no replicas will result in a *yellow* cluster health. If you
```
ah thx for the correction. I copied this from the original text

reworded:

> If you only have one node, CrateDB will not create any replicas. If you have more than one node, CreateDB will create one replica per shard.
>
> This range is the default value.

norosa commented on docs/general/ddl/replication.rst:
```diff
@@ -1,73 +1,85 @@
+``2-4``   Each table will require at least two replicas for CrateDB to consider
+          it fully replicated (i.e., a *green status*).
+
+          If the cluster has four nodes, CrateDB will create four replicas,
```
oops good catch! thx

reworded to fix. I also removed the now redundant note about resiliency as I cover this explicitly in the intro (which is where such an important point belongs imo)

> If the cluster has five nodes, CrateDB will create four replicas, with each replica located on a different node from its respective primary.
>
> If a cluster has four or fewer nodes, CreateDB will have to locate one or more replica shards will be on the same node as the respective primary shard. If this happens, your cluster will have a yellow cluster health.

norosa commented on docs/general/ddl/replication.rst:
```diff
@@ -1,73 +1,85 @@
+          If a cluster has three or fewer nodes, CreateDB will have to locate one
+          or more  replica shards will be on the same node as the respective
+          primary shard. If this happens, your cluster will have a *yellow*
+          cluster health.
```
fixed as above

norosa commented on docs/general/ddl/replication.rst:
```diff
@@ -1,73 +1,85 @@
+          primary shard. If this happens, your cluster will have a *yellow*
+          cluster health.
+--------- ---------------------------------------------------------------------
+``0-all`` CrateDB will create as many replicas as you have nodes available.
```
I want to avoid the inline literals if possible. how about this:

> CrateDB will create one replica shard for every node that is available in addition to the node that holds the primary shard.

norosa commented:
@seut thx so much for the review! pushed a fixup with the same changes as commented. left the comments all open for you to resolve if you like the changes

seut commented on docs/general/ddl/replication.rst:
```diff
@@ -1,73 +1,100 @@
-.. contents::
-    :local:
+CrateDB :ref:`allocates <gloss-shard-allocation>` every shard to a specific
+node. CrateDB will try to allocate the primary shard and replica shards to
+different nodes to maximize :ref:`durability <concept-durability>` and
+:ref:`resiliency <concept-resiliency>`.
```
```suggestion
node. CrateDB will enforce the allocation of the primary shard and replica 
shards on different nodes to maximize :ref:`durability <concept-durability>`
and :ref:`resiliency <concept-resiliency>`.
If not enough nodes are available for each replica shard to be allocated,
the table will go into a `YELLOW` health state, see :ref:`sys.health <sys-health>`.
```

norosa replied:
ty!!

I reworded slightly::

    CrateDB :ref:`allocates <gloss-shard-allocation>` every shard to a specific
    node. Normally, CrateDB will enforce the requirement that the primary shard and
    replica shards are each allocated to a different node. If that requirement
    cannot be satisfied, the cluster will be given a *yellow* :ref:`health status
    <sys-health>`.

norosa replied:
gotcha. thanks!

seut commented on docs/general/ddl/replication.rst:
```diff
@@ -1,73 +1,100 @@
-.. NOTE::
-
-    The number of replicas can be changed at any time.
+As well as configuring a single fixed number of pre-replica shards, you can use
```
What do you mean with `pre-replica shards`? 

```suggestion
As well as configuring a single fixed number of replica shards (shard copies), you can use
```

norosa replied:
oops! good catch thanks. typo. I think I meant "per-shard replicas" ("per" not "pre" and mixed up the nouns)

I reworded:

> As well as being able to configure a fixed number of replicas, you can configure a range of values by using a string to specify a minimum and a maximum (dependent on the number of nodes in the cluster).

seut commented on docs/general/ddl/replication.rst:
```diff
@@ -1,73 +1,100 @@
+          with each replica located on a different node from its respective
+          primary.
+
+          If a cluster has four or fewer nodes, CreateDB will have to locate
+          one or more replica shards will be on the same node as the respective
+          primary shard. If this happens, your cluster will have a *yellow*
```
```suggestion
          If a cluster has four or fewer nodes, CreateDB would have to locate
          one or more replica shards on the same node as the respective
          primary shard. This is not permitted and such your cluster will have a *yellow*
```

seut replied:
> because even though it's not permitted, it can still happen

No this cannot happen, its really enforced. CrateDB will never allocate a replica shard on the same node of the primary shard.

seut replied:
So we should also remove the "Normally" inside `Normally, CrateDB dynamically allocates shards to...`

norosa replied:
ah, I see. thanks for the explanation! I have an outstanding question for you that is related to this. see https://github.com/crate/crate/issues/11263

if you have time, could you take a look at that? based on the discussion there, I think I should probably make a follow-up PR to really make this behavior clear in other places too

seut reviewed (commented):
Thanks for the corrections, but some things are still not correct. Added suggestions.

norosa commented:
@seut thanks for the second review! I pushed another fix up. quite happy with the improvements to `replication.rst` now! thanks for helping me shape it up

norosa commented:
@seut pushed a new fixup

thanks so much for your continued help with this. my confusion makes me think that we need a new section somewhere (perhaps on an existing page) that clearly describes how allocation works (as in, all the cool ways CrateDB can or will try to optimize allocation for you) -- thoughts?

https://github.com/crate/crate/issues/11263 (linked in a prev comment) is a good example of what I mean. this should be in the docs imo

---

I had to slash most of the changes I made before pushing this fixup. for example, I spent a good chunk of time trying to explain things like:

- how does CrateDB allocate shards so as to spread out primary and replica workloads for clusters with fewer nodes than the total shards (primary and replica)? does CrateDB do this?

I gave up and decided it was out of scope for this PR. I have made a note to revisit this later

to give you an idea, here's something I cut:

```
To illustrate this, consider the following hypothetical allocations for a two
node cluster with one table having two shards, each with one replica:

+-------------------------------+---------------------------------------------+
| Allocations                   | Comment                                     |
+===============================+=============================================+
| +-------------+-------------+ | This arrangement is suboptimal.             |
| | Node 1      | Node 2      | |                                             |
| +=============+=============+ | Table data can be can be fully recovered if |
| | Primary A   | Replica A   | | a single node is lost. However, primary and |
| +-------------+-------------+ | replication workloads are not shared evenly |
| | Primary B   | Replica B   | | between the nodes.                          |
| +-------------+-------------+ |                                             |
+-------------------------------+---------------------------------------------+
| +-------------+-------------+ | This arrangement is optimal.                |
| | Node 1      | Node 2      | |                                             |
| +=============+=============+ | Table data can be can be fully recovered if |
| | Primary A   | Primary B   | | a single node lost and primary and          |
| +-------------+-------------+ | replication workloads are shared evenly     |
| | Replica B   | Replica A   | | between nodes.                              |
| +-------------+-------------+ |                                             |
+-------------------------------+---------------------------------------------+
| +-------------+-------------+ | This arrangement is not permitted.          |
| | Node 1      | Node 2      | |                                             |
| +=============+=============+ | Without :ref:`backups <snapshot-restore>`,  |
| | Primary A   | Primary B   | | the loss of either node would result in     |
| +-------------+-------------+ | permanent data loss (i.e., both primary and |
| | Replica A   | Replica B   | | replica shard).                             |
| +-------------+-------------+ |                                             |
+-------------------------------+---------------------------------------------+
```

(my explanation might be wrong. but hopefully, gives an idea of where I thought I was going with this)

norosa commented:
note to self: there's "normally," that still needs removing

seut commented:
@norosa I think, publishing some kind of sharding and replication concept descriptions and some graphics to illustrate that better is a great idea. Maybe it would best fit into the `Concepts` section of the CrateDB reference documentation. But definitely a different, maybe larger task :)

norosa commented:
I ran into difficulties rebasing this with the automatic merge commit GitHub added from master. not sure why, but too many merge conflicts for me to be comfortable with. instead of wasting time trying to get this branch fixed, I recreated my changes for a new PR: https://github.com/crate/crate/pull/11360

norosa commented:
@seut I have reopened this pr for you

the changes I made yesterday are here:
 on this PR: https://github.com/crate/crate/pull/11319/commits/55518bde2babb4c2cd9dd3d87d43e7cb2a8c415c

(after pushing this commit I had issues rebasing. I think it's because I did a merge from master using the github interface and my local client is having a hard time reconciling the history. I will have another go at sorting it out once you're happy with my changes on this branch)

summary of [commentary I left on the other pr](https://github.com/crate/crate/pull/11360#issuecomment-833777431:) (which I have closed) 

- moved "Table health" down the page and split it into two short sections
- "Shard recovery" goes over the BASICS of _how_ replicas are used. I briefly alluded to CAP. we can come back to expand our coverage of this topic later (as discussed)
  - this is a useful section imo because it makes it clear why more replicas and more thinly distributed shards are beneficial. it also makes it easy to explain underreplication (next item)
- "Underreplication" explains (by way of example) why a single node cannot hold the primary and replica of the same shard and uses more descriptive terminology ("underreplication" vs "yellow health")

my question was:

- can a single node hold two replicas of the same shard? by my logic, one of those replicas would be useless (like with the primary and replica on the same node). if the answer is "yes", would it be helpful to mention that and explain why it is allowed? if not, should we make the rule more generic, i.e., "no node can hold two or more copies (primary or replica) of the same shard

seut commented:
>  can a single node hold two replicas of the same shard?

No this is not allowed as it wouldn't make much sense. So yes "no node can hold two or more copies (primary or replica) of the same shard".

seut commented on docs/general/ddl/replication.rst:
```diff
@@ -110,6 +95,86 @@ time.
+(i.e., fewer shards per node) can both increase chances of a :ref:`successful
+recovery <ddl-replication-recovery>` in the event of node loss.
+
+Although not ideal, a single node can hold multiple shards belonging to the
+same table. For example, suppose a table has more shards (primaries and
```
Having multiple shards of the same table on one node is a common pattern and even recommended as it increases parallel execution per read query (multiple threads can be used).

norosa replied:
gotcha thx!

norosa commented:
@seut thanks for your clarifications! pushed a fixup and included a second example

seut reviewed (approved):
👍 Thanks for all the effort.
